src/encoding.rs

```diff
@@ -870,23 +850,86 @@ pub mod string {
     }
 
     length_delimited!(String);
+
+    #[cfg(test)]
+    mod test {
+        use quickcheck::{quickcheck, TestResult};
+
+        use super::super::test::{check_collection_type, check_type};
+        use super::*;
+
+        quickcheck! {
+            fn check(value: String, tag: u32) -> TestResult {
+                super::test::check_type(value, tag, WireType::LengthDelimited,
+                                        encode, merge, encoded_len)
+            }
+            fn check_repeated(value: Vec<String>, tag: u32) -> TestResult {
+                super::test::check_collection_type(value, tag, WireType::LengthDelimited,
+                                                   encode_repeated, merge_repeated,
+                                                   encoded_len_repeated)
+            }
+        }
+    }
+}
+
+pub trait BytesAdapter: Default + Sized + 'static {
```

Comment:
I'm going to poke at this a bit, I think it should be possible.

Reply:
Well this at least compiles/passes tests:

Reply:
will keep playing, the most important thing is to make sure this doesn't regress the performance of the existing vec impl.

Reply:
After looking at this more, the encoding APIs get tricky to impl because `Vec<u8>` does not impl `Buf`.  I think it may be possible to have two separate implementations, one for `Vec<u8>` and one that is generic for anything which impls `Buf`.

Reply:
Indeed this is not possible because `Vec<u8>` is not a `Buf`. I have now changed the trait to be private (with a workaround stolen from https://docs.rs/tokio/0.2.21/tokio/net/trait.ToSocketAddrs.html). Does that address your concern?

Otherwise I guess it would be necessary to create a new function, say, `encode_buf(...)` and a lot of special-casing in `prost-derive` for `Bytes`-backed `bytes` fields... That seems like a much less elegant option.